packages/router/router.ts

```diff
@@ -3929,41 +4084,136 @@ async function loadLazyRouteModule(
+            // of whether exposing the current data to the user is useful?
+            Promise.resolve({ type: ResultType.data, result: undefined });
+      };
+
+      return {
+        ...match,
+        shouldLoad,
+        resolve,
+      };
+    }),
+    request,
+    params: matches[0].params,
+  });
+
+  // Throw if any loadRoute implementations not called since they are what
+  // ensures a route is fully loaded
+  // Throw if any loadRoute implementations not called since they are what
+  // ensure a route is fully loaded
+  matches.forEach((m) =>
+    invariant(
+      loadedMatches.has(m.route.id),
+      `\`match.resolve()\` was not called for route id "${m.route.id}". ` +
+        "You must call `match.resolve()` on every match passed to " +
+        "`dataStrategy` to ensure all routes are properly loaded."
+    )
+  );
+
+  // Filter out any middleware-only matches for which we didn't need to run handlers
+  return results.filter((_, i) => routeIdsToLoad.has(matches[i].route.id));
 }
 
+// Default logic for calling a loader/action is the user has no specified a dataStrategy
 async function callLoaderOrAction(
   type: "loader" | "action",
   request: Request,
   match: AgnosticDataRouteMatch,
-  matches: AgnosticDataRouteMatch[],
   manifest: RouteManifest,
   mapRouteProperties: MapRoutePropertiesFunction,
-  basename: string,
-  v7_relativeSplatPath: boolean,
-  opts: {
-    isStaticRequest?: boolean;
-    isRouteRequest?: boolean;
-    requestContext?: unknown;
-  } = {}
-): Promise<DataResult> {
-  let resultType;
+  handlerOverride: Parameters<DataStrategyMatch["resolve"]>[0],
+  staticContext?: unknown
+): Promise<HandlerResult> {
   let result;
   let onReject: (() => void) | undefined;
 
-  let runHandler = (handler: ActionFunction | LoaderFunction) => {
+  let runHandler = (
+    handler: AgnosticRouteObject["loader"] | AgnosticRouteObject["action"]
+  ) => {
     // Setup a promise we can race against so that abort signals short circuit
     let reject: () => void;
     let abortPromise = new Promise((_, r) => (reject = r));
     onReject = () => reject();
     request.signal.addEventListener("abort", onReject);
-    return Promise.race([
-      handler({
-        request,
-        params: match.params,
-        context: opts.requestContext,
-      }),
-      abortPromise,
-    ]);
+
+    let runHandlerForReal = (ctx?: unknown) => {
+      if (typeof handler !== "function") {
+        return Promise.reject(
+          new Error(
+            `You cannot call the handler for a route which defines a boolean ` +
+              `"${type}" [routeId: ${match.route.id}]`
+          )
+        );
+      }
+      return handler(
+        {
+          request,
+          params: match.params,
+          context: staticContext,
+        },
+        ...(ctx !== undefined ? [ctx] : [])
+      );
+    };
+
+    let handlerPromise = handlerOverride
+      ? handlerOverride(async (ctx: unknown) => runHandlerForReal(ctx))
```

Comment:
If the user gave us a `handlerOverride` then we call that with a function letting them proxy a second argument into the loader/action